Sources/PackageGraph/RegistryPackageContainer.swift

```diff
@@ -0,0 +1,193 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import PackageLoading
+import PackageModel
+import PackageRegistry
+import SourceControl
+
+import TSCBasic
+import TSCUtility
+
+import Dispatch
+
+/// Adaptor to expose an individual repository as a package container.
+public final class RegistryPackageContainer: BasePackageContainer {
+    /// This is used to remember if tools version of a particular version is
+    /// valid or not.
+    public private(set) var validToolsVersionsCache: [Version: Bool] = [:]
+
+    /// The available version list (in reverse order).
+    override public func versions(filter isIncluded: (Version) -> Bool) -> AnySequence<Version> {
+        return AnySequence(self._reversedVersions.filter(isIncluded).lazy.filter {
+            // If we have the result cached, return that.
+            if let result = self.validToolsVersionsCache[$0] {
+                return result
+            }
+
+            // Otherwise, compute and cache the result.
+            let isValid = (try? self.toolsVersion(for: $0)).flatMap(self.isValidToolsVersion(_:)) ?? false
+            self.validToolsVersionsCache[$0] = isValid
+            return isValid
+        })
+    }
+
+    override public var reversedVersions: [Version] { self._reversedVersions }
+
+    /// The versions in the repository sorted by latest first.
+    let _reversedVersions: [Version]
+
+    let registryManager: RegistryManager
+
+    private var manifestCache: [Version: Manifest] = [:]
+
+    init(
+        identifier: PackageReference,
+        mirrors: DependencyMirrors,
+        registryManager: RegistryManager,
+        manifestLoader: ManifestLoaderProtocol,
+        toolsVersionLoader: ToolsVersionLoaderProtocol,
+        currentToolsVersion: ToolsVersion
+    ) throws {
+        self.registryManager = registryManager
+
+        let versionsWithDuplicates = try await { registryManager.fetchVersions(of: identifier, completion: $0) }
```

Comment:
if I read this correctly this would block on a network call, is that true?

Reply:
Yes, that's right. This is a shortcoming of the existing abstraction provided by `BasePackageContainer`. What do you think? Should we recolor that base class to make everything asynchronous or is it acceptable to block (given a reasonable timeout)?